# 2020.04.06 (Mon)

## Starting the annotations from scratch (`RF sentences`)

Following the official guide:
- https://www.youtube.com/watch?v=59BKHO_xBPA
- https://github.com/explosion/projects/tree/master/ner-food-ingredients#data-creation-and-training-workflow

### Label data manually

In [2]:
!prodigy ner.manual cord_19_rf_sentences blank:en data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR

Using 1 label(s): RISK_FACTOR

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!

^C

✔ Saved 102 annotations to database SQLite
Dataset: cord_19_rf_sentences
Session ID: 2020-04-06_22-08-56



### Pretrain different models

#### `en_vectors_web_lg`

In [5]:
!prodigy train ner cord_19_rf_sentences en_vectors_web_lg --eval-split 0.2

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     389.18       0.000      0.000      0.000                                 
 2     191.03      25.000      4.000      6.897                                 
 3     265.45      35.294     24.000     28.571                                 
 4     246.57      28.571     24.000     26.087                                 
 5     239.21      22.727     20.000     21.277                                 
 6     215.80      42.857     24.000     30.769                                 
 7     258.40      42.105     32.000     36.364                                 
 8     237.19      50.000     32.000   

#### `en_core_web_lg`

In [6]:
!prodigy train ner cord_19_rf_sentences en_core_web_lg --eval-split 0.2

✔ Loaded model 'en_core_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1421.81       0.000      0.000      0.000                                 
 2    1245.16       0.000      0.000      0.000                                 
 3    1262.77       0.000      0.000      0.000                                 
 4    1249.04       0.000      0.000      0.000                                 
 5    1262.46       0.000      0.000      0.000                                 
 6    1205.26       0.000      0.000      0.000                                 
 7    1137.92       0.000      0.000      0.000                                 
 8    1131.84       0.000      0.000      

⁉️ Why is `en_core_web_lg` producing no results at all?

#### `en_core_sci_lg`

In [14]:
!prodigy train ner cord_19_rf_sentences en_core_sci_lg --eval-split 0.2 --output models/2020_04_06_rf_sentences

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     559.17       0.000      0.000      0.000                                 
 2     550.33      33.333      8.000     12.903                                 
 3     538.43      50.000     24.000     32.432                                 
 4     457.89      52.632     40.000     45.455                                 
 5     452.06      50.000     36.000     41.860                                 
 6     415.01      60.000     48.000     53.333                                 
 7     489.45      47.826     44.000     45.833                                 
 8     461.13      45.000     36.000     4

#### `en_vectors_web_lg` + tok2vec

In [8]:
!prodigy train ner cord_19_rf_sentences en_vectors_web_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_vectors_model436.bin

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_vectors_model436.bin
embed_rows: 2000 | require_vectors: True | cnn_maxout_pieces: 3 |
token_vector_width: 96 | conv_depth: 4 | nr_feature_tokens: 3 |
pretrained_vectors: en_vectors_web_lg.vectors
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     339.88       0.000      0.000      0.000                                 
 2     199.85      50.000     16.000     24.242                                 
 3     128.26      50.000     28.000     35.897                                 
 4      76.43      42.857     24.000     30.769                                 
 5      63.28      60.000     3

#### `en_core_web_lg` + tok2vec

In [9]:
!prodigy train ner cord_19_rf_sentences en_core_web_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_vectors_model436.bin

✔ Loaded model 'en_core_web_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_vectors_model436.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2055.18       0.000      0.000      0.000                                 
 2    1808.48       0.000      0.000      0.000                                 
 3    1624.76       0.000      0.000      0.000                                 
 4    1443.31       0.000      0.000      0.000                                 
 5    1652.93       0.000      0.000      0.000                                 
 6    1470.22       0.000      0.000      0.000                                 
 7    1345.14       0.000      0.000      0.000

#### `en_core_sci_lg` + tok2vec

In [13]:
!prodigy train ner cord_19_rf_sentences en_core_sci_lg --eval-split 0.2 \
    --init-tok2vec models/tok2vec_rf_sci_model328.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 10
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model328.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     536.68       0.000      0.000      0.000                                 
 2     631.61       0.000      0.000      0.000                                 
 3     539.99      40.000     24.000     30.000                                 
 4     505.27      47.619     40.000     43.478                                 
 5     508.05      45.000     36.000     40.000                                 
 6     461.69      52.941     36.000     42.857                                 
 7     473.25      45.000     36.000     40.000    

🔥 Top model so far: `53.333` 🔥

### Label more data by correcting the model's predictions

In [15]:
!prodigy ner.correct cord_19_rf_sentences_correct models/2020_04_06_rf_sentences data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR --exclude cord_19_rf_sentences

Using 1 label(s): RISK_FACTOR
Added dataset cord_19_rf_sentences_correct to database SQLite.

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!

^C

✔ Saved 220 annotations to database SQLite
Dataset: cord_19_rf_sentences_correct
Session ID: 2020-04-06_23-58-48



### Train the final model

For futher training, only `en_core_sci_lg` will be used.

In [3]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1554.25      61.538     37.500     46.602                                 
 2    1346.09      64.407     59.375     61.789                                 
 3    1407.15      66.667     50.000     57.143                                 
 4    1411.73      72.727     50.000     59.259                                 
 5    1412.81      73.913     53.125     61.818                                 
 6    1319.96      73.913     53.125     61.818                                 
 7    1378.05      72.917     54.688     62.500                                 
 8    1208.99      72.727     62.500    

In [6]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model328.bin --output models/2020_04_07_rf_sentences_corrected

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model328.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1578.16      64.000     25.000     35.955                                 
 2    1439.61      55.319     40.625     46.847                                 
 3    1528.20      72.222     60.938     66.102                                 
 4    1429.18      71.154     57.812     63.793                                 
 5    1414.46      69.388     53.125     60.177                                 
 6    1344.69      75.472     62.500     68.376                                 
 7    1472.62      70.909     60.938     65.546  

In [5]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model989.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model989.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1570.28      62.500     23.438     34.091                                 
 2    1397.40      59.091     40.625     48.148                                 
 3    1485.47      70.833     53.125     60.714                                 
 4    1445.56      68.293     43.750     53.333                                 
 5    1431.31      64.000     50.000     56.140                                 
 6    1339.87      70.833     53.125     60.714                                 
 7    1387.07      72.340     53.125     61.261  

# 2020.04.09 (Thu)
## Experiments with new `tok2vec` models

In [1]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model978_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1577.68      57.895     34.375     43.137                                 
 2    1386.95      67.308     54.688     60.345                                 
 3    1455.49      76.596     56.250     64.865                                 
 4    1457.04      79.592     60.938     69.027                                 
 5    1413.30      78.431     62.500     69.565                                 
 6    1351.85      76.923     62.500     68.966                                 
 7    1304.38      78.000     60.938     68.4

In [2]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_rf_sci_model997_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model997_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1630.03      55.556     23.438     32.967                                 
 2    1442.41      55.556     46.875     50.847                                 
 3    1429.19      65.957     48.438     55.856                                 
 4    1467.20      65.455     56.250     60.504                                 
 5    1467.20      69.643     60.938     65.000                                 
 6    1376.49      67.241     60.938     63.934                                 
 7    1324.06      73.585     60.938     66.6

In [3]:
!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg --eval-split 0.2 --n-iter 20 \
    --init-tok2vec models/tok2vec_abs_fil_sci_model975_gpu.bin

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1587.95      62.963     26.562     37.363                                 
 2    1500.37      60.784     48.438     53.913                                 
 3    1510.36      63.492     62.500     62.992                                 
 4    1519.20      60.000     56.250     58.065                                 
 5    1521.26      65.574     62.500     64.000                                 
 6    1495.73      64.912     57.812     61.157                                 
 7    1386.50      65.517     59.375    

# 2020.04.10 (Fri)
## Experiments with new `tok2vec` models (cont.)

The Kaggle notebook training models for the full set of abstracts (i.e. `cord_19_abstracts.jsonl`) crashed, so I have only partial results and no loss metrics for them.

In [11]:
import os

from pathlib import Path

# run training for a selected models, use every skip_models-th file
skip_models = 13
abs_models_dir = Path('models/tok2vec_abs_sci_ALL_gpu')

models = sorted(os.listdir(abs_models_dir))

for idx, model_file in enumerate(models):
    if idx % skip_models == 0:
        model_path = abs_models_dir / model_file
        !prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg \
            --eval-split 0.2 --n-iter 20 --init-tok2vec $model_path

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model100.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1570.29      43.478     31.250     36.364                                 
 2    1421.36      59.574     43.750     50.450                                 
 3    1486.19      68.519     57.812     62.712                                 
 4    1477.14      70.000     54.688     61.404                                 
 5    1389.42      73.469     56.250     63.717                                 
 6    1377.87      70.000     54.688     61.404                                 
 7    1379.27      65.385     53.125    

15    1143.57      69.231     56.250     62.069                                 
16    1108.60      69.091     59.375     63.866                                 
17    1037.24      66.667     59.375     62.810                                 
18    1059.90      67.857     59.375     63.333                                 
19    1029.31      65.517     59.375     62.295                                 
20    1091.05      67.797     62.500     65.041                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      76.364   65.625    70.588


Best F-Score   70.588
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs

✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model230.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1537.49      65.714     35.938     46.465                                 
 2    1506.48      51.064     37.500     43.243                                 
 3    1436.09      68.627     54.688     60.870                                 
 4    1438.86      67.925     56.250     61.538                                 
 5    1471.17      58.730     57.812     58.268                                 
 6    1395.94      76.786     67.188     71.667                                 
 7    1329.15      71.186     65.625     68.293                                 
 8    1224.20      70.000     65.625     67.742                                 
 9    1212.04      72.222     60.938     66.102                     

17    1045.76      74.545     64.062     68.908                                 
18    1047.72      74.545     64.062     68.908                                 
19    1039.92      76.786     67.188     71.667                                 
20    1076.51      75.000     65.625     70.000                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      81.818   70.312    75.630


Best F-Score   75.630
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_ALL_gpu/model321.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1483.42      63.415     40.625     49.524                                 
 2    1529.89      60.000     60.938     60.465                                 
 3    1444.98      66.154     67.188     66.667                                 
 4    1448.20      68.421     60.938     64.463                                 
 5    1388.45      67.692     68.750     68.217                                 
 6    1309.81      65.574     62.500     64.000                                 
 7    1318.09      68.852     65.625     67.200                                 
 8    1223.92      66.667     65.625     66.142                                 
 9    1219.50      68.254     67.188     67.717                                 
10    1226.45      68.182     70.312     69.231                     

### Training final models with selected `tok2vec` models

In [3]:
import os

from pathlib import Path


def train_ner_with_every_tok2vec(datasets, base_model="en_core_sci_lg"):
    tok2vec_dir = Path("models")
    tok2vecs = sorted(os.listdir(tok2vec_dir))
    
    for tok2vec_file in tok2vecs:
        if "_sci_" not in tok2vec_file:
            continue

        tok2vec_path = tok2vec_dir / tok2vec_file
        !prodigy train ner $datasets $base_model --eval-split 0.2 --n-iter 20 \
            --init-tok2vec $tok2vec_path

In [6]:
train_ner_with_every_tok2vec("cord_19_rf_sentences,cord_19_rf_sentences_correct")

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1587.95      62.963     26.562     37.363                                 
 2    1500.37      60.784     48.438     53.913                                 
 3    1510.36      63.492     62.500     62.992                                 
 4    1519.20      60.000     56.250     58.065                                 
 5    1521.26      65.574     62.500     64.000                                 
 6    1495.73      64.912     57.812     61.157                                 
 7    1386.50      65.517     59.375    

16    1146.76      78.431     62.500     69.565                                 
17    1045.76      74.545     64.062     68.908                                 
18    1047.72      74.545     64.062     68.908                                 
19    1039.92      76.786     67.188     71.667                                 
20    1076.51      75.000     65.625     70.000                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      81.818   70.312    75.630


Best F-Score   75.630
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model336_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨ 

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1577.68      57.895     34.375     43.137                                 
 2    1386.95      67.308     54.688     60.345                                 
 3    1455.49      76.596     56.250     64.865                                 
 4    1457.04      79.592     60.938     69.027                                 
 5    1413.30      78.431     62.500     69.565                                 
 6    1351.85      76.923     62.500     68.966                                 
 7    1304.38      78.000     60.938     68.421                                 
 8    1255.67      76.471     60.938     67.826                                 
 9    1219.66      78.431     62.500     69.565                                 
10    1203.83      74.545     64.062     68.908                     

🔥 Top model so far: `75.630` 🔥
- trained with `tok2vec_abs_sci_model308_gpu.bin`

In [15]:
#!prodigy train ner cord_19_rf_sentences,cord_19_rf_sentences_correct en_core_sci_lg \
#    --eval-split 0.2 --n-iter 20 --init-tok2vec models/tok2vec_abs_sci_model308_gpu.bin \
#    --output models/2020_04_10_en_ner_rf_sm

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model308_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1518.02      61.905     40.625     49.057                                 
 2    1389.84      56.897     51.562     54.098                                 
 3    1485.12      71.698     59.375     64.957                                 
 4    1385.46      70.175     62.500     66.116                                 
 5    1455.78      72.727     62.500     67.227                                 
 6    1404.07      78.947     70.312     74.380                                 
 7    1323.94      81.818     70.312     75.

# 2020.04.13 (Mon)
## Label more data by correcting the model's predictions (part 2)

In [ ]:
!prodigy ner.correct cord_19_rf_sentences_correct_2 models/2020_04_10_en_ner_rf_sm data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR --exclude cord_19_rf_sentences,cord_19_rf_sentences_correct

Using 1 label(s): RISK_FACTOR
Added dataset cord_19_rf_sentences_correct_2 to database SQLite.

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!



## Train final models, with 3 datasets

In [7]:
train_ner_with_every_tok2vec("cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2")

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2885.30      59.302     42.149     49.275                                 
 2    2552.55      62.353     43.802     51.456                                 
 3    2728.96      65.385     56.198     60.444                                 
 4    2487.82      72.917     57.851     64.516                                 
 5    2373.22      72.642     63.636     67.841                                 
 6    2348.38      74.074     66.116     69.869                                 
 7    2218.46      73.832     65.289   

16    1818.28      75.652     71.901     73.729                                 
17    1795.52      75.000     71.901     73.418                                 
18    1827.94      74.561     70.248     72.340                                 
19    1761.02      74.783     71.074     72.881                                 
20    1806.65      73.504     71.074     72.269                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      77.193   72.727    74.894


Best F-Score   74.894
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model336_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨

✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2835.29      68.116     38.843     49.474                                 
 2    2732.17      72.500     47.934     57.711                                 
 3    2570.86      78.000     64.463     70.588                                 
 4    2535.81      72.381     62.810     67.257                                 
 5    2355.29      78.641     66.942     72.321                                 
 6    2273.33      73.333     63.636     68.142                                 
 7    2248.22      75.472     66.116     70.485                                 
 8    2121.12      72.477     65.289     68.696                                 
 9    2055.64      72.642     63.636     67.841                          

🔥 More annotations lowered the performance; top model: `74.894` 🔥
- trained with `tok2vec_abs_sci_model308_gpu.bin`

### Check if `sci` is still the best choice

In [3]:
datasets_0_1_2 = "cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2"
model_0_1_2 = "models/2020_04_13_en_ner_rf_sm"

In [10]:
for base_model in ['en_vectors_web_lg', 'en_core_web_lg', 'en_core_sci_lg']:
    !prodigy train ner $datasets_0_1_2 $base_model --eval-split 0.2 --n-iter 20

✔ Loaded model 'en_vectors_web_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1517.43      63.793     30.579     41.341                                 
 2    1373.26      68.750     45.455     54.726                                 
 3     997.71      69.048     47.934     56.585                                 
 4     903.18      71.910     52.893     60.952                                 
 5     663.74      75.269     57.851     65.421                                 
 6     690.64      74.118     52.066     61.165                                 
 7     499.83      75.862     54.545     63.462                                 
 8     439.64      73.118     56.198

`en_core_sci_lg` is still the best base model, although the margins are getting smaller.

In [2]:
!prodigy train ner $datasets_0_1_2 en_core_sci_lg \
    --output $model_0_1_2 \
    --init-tok2vec models/tok2vec_abs_sci_model308_gpu.bin \
    --eval-split 0.2 --n-iter 20

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model308_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2811.45      69.231     52.066     59.434                                 
 2    2606.03      67.961     57.851     62.500                                 
 3    2571.19      73.529     61.983     67.265                                 
 4    2442.88      75.238     65.289     69.912                                 
 5    2329.50      70.690     67.769     69.198                                 
 6    2213.83      77.273     70.248     73.593                                 
 7    2127.51      75.926     67.769     71

## Label more data by correcting the model's predictions (part 3)

In [4]:
!prodigy ner.correct cord_19_rf_sentences_correct_3 $model_0_1_2 data/raw/cord_19_rf_sentences.jsonl \
    --label RISK_FACTOR --exclude $datasets_0_1_2

Using 1 label(s): RISK_FACTOR
Added dataset cord_19_rf_sentences_correct_3 to database SQLite.

✨  Starting the web server at http://localhost:8080 ...
Open the app in your browser and start annotating!

^C

✔ Saved 79 annotations to database SQLite
Dataset: cord_19_rf_sentences_correct_3
Session ID: 2020-04-13_23-22-28



## Train final models, with 4 datasets

In [10]:
datasets_0_1_2_3 = "cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2," \
    "cord_19_rf_sentences_correct_3"
model_0_1_2_3 = "models/2020_04_14_en_ner_rf_sm"

In [13]:
train_ner_with_every_tok2vec(datasets_0_1_2_3)

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 625 total examples
Using 500 train / 125 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    3381.36      59.854     51.899     55.593                                 
 2    3083.86      62.420     62.025     62.222                                 
 3    2890.69      61.585     63.924     62.733                                 
 4    2752.00      60.209     72.785     65.903                                 
 5    2699.75      61.850     67.722     64.653                                 
 6    2627.96      62.428     68.354     65.257                                 
 7    2529.31      62.805     65.190   

16    2240.57      61.272     67.089     64.048                                 
17    2205.91      60.819     65.823     63.222                                 
18    2198.57      62.573     67.722     65.046                                 
19    2103.46      64.497     68.987     66.667                                 
20    2157.28      64.497     68.987     66.667                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      64.740   70.886    67.674


Best F-Score   67.674
Baseline       0.000              

✔ Loaded model 'en_core_sci_lg'
Created and merged data for 625 total examples
Using 500 train / 125 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model336_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨

✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    3342.47      50.595     53.797     52.147                                 
 2    3156.18      61.006     61.392     61.199                                 
 3    2907.23      61.111     62.658     61.875                                 
 4    2676.68      63.580     65.190     64.375                                 
 5    2700.87      62.428     68.354     65.257                                 
 6    2693.58      66.272     70.886     68.502                                 
 7    2495.51      65.854     68.354     67.081                                 
 8    2502.03      64.671     68.354     66.462                                 
 9    2305.59      63.793     70.253     66.867                          

🔥 Top model for a full set of annotations: `68.770` 🔥
- trained with `tok2vec_abs_sci_model304_gpu.bin`

**Why such a significant performance drop?!?**

## Export the annotated datasets

In [28]:
datasets_merged = "cord_19_rf_sentences_merged"
file_db_out = f"data/annotated/{datasets_merged}.jsonl"
file_spacy_format = f"data/annotated/{datasets_merged}.json"

In [16]:
!prodigy db-merge $datasets_0_1_2_3 $datasets_merged

✔ Created dataset 'cord_19_rf_sentences_merged'
✔ Merged 652 examples from 4 datasets
Created merged dataset 'cord_19_rf_sentences_merged'


In [29]:
!prodigy db-out $datasets_merged > data/annotated/cord_19_rf_sentences_merged.jsonl

In [30]:
!prodigy data-to-spacy data/annotated/cord_19_rf_sentences_merged.json --ner $datasets_merged

ℹ Starting with language en
Created and merged data for 625 total examples

Type   Total   Merged
----   -----   ------
NER      652      625

✔ Saved 625 examples to
data/annotated/cord_19_rf_sentences_merged.json


# 2020.04.14 (Tue)
## Train final models with `md` base model

In [6]:
datasets_list = [
    "cord_19_rf_sentences",
    "cord_19_rf_sentences,cord_19_rf_sentences_correct",
    "cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2",
    "cord_19_rf_sentences,cord_19_rf_sentences_correct,cord_19_rf_sentences_correct_2,cord_19_rf_sentences_correct_3"
]
for datasets in datasets_list:
    print(f"= {datasets} ==================================")
    train_ner_with_every_tok2vec(datasets, "en_core_sci_md")

= cord_19_rf_sentences ==================================
✔ Loaded model 'en_core_sci_md'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_fil_sci_model975_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     473.65       0.000      0.000      0.000                                 
 2     717.18       0.000      0.000      0.000                                 
 3     562.82      26.316     20.000     22.727                                 
 4     539.94      31.250     20.000     24.390                                 
 5     490.88      40.000     24.000     30.000                                 
 6     427.17      43.750     28.000     34.146                  

 4     519.65      47.368     36.000     40.909                                 
 5     510.44      38.889     28.000     32.558                                 
 6     444.14      42.105     32.000     36.364                                 
 7     662.63      40.000     40.000     40.000                                 
 8     469.91      31.818     28.000     29.787                                 
 9     484.98      38.095     32.000     34.783                                 
10     445.55      52.381     44.000     47.826                                 
11     457.62      47.826     44.000     45.833                                 
12     414.88      50.000     44.000     46.809                                 
13     403.09      57.143     48.000     52.174                                 
14     420.27      60.000     48.000     53.333                                 
15     441.94      60.000     48.000     53.333                                 
16     436.54      60.000   

14     402.47      50.000     40.000     44.444                                 
15     442.71      50.000     40.000     44.444                                 
16     415.41      50.000     36.000     41.860                                 
17     356.17      55.556     40.000     46.512                                 
18     355.91      66.667     48.000     55.814                                 
19     348.44      66.667     48.000     55.814                                 
20     363.66      61.111     44.000     51.163                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      66.667   48.000    55.814


Best F-Score   55.814
Baseline       0.000              

✔ Loaded model 'en_core_sci_md'
Created and merged data for 99 total examples
Using 80 train / 19 eval (split 20%)
Component: ner | Batch size: compounding | Dropo

✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model978_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1     468.97       0.000      0.000      0.000                                 
 2     586.66       0.000      0.000      0.000                                 
 3     537.96      50.000     32.000     39.024                                 
 4     543.50      44.444     32.000     37.209                                 
 5     475.28      42.857     36.000     39.130                                 
 6     467.28      41.176     28.000     33.333                                 
 7     463.37      55.000     44.000     48.889                                 
 8     499.87      40.000     40.000     40.000                                 
 9     446.57      42.308     44.000     43.137                          

 7    1345.34      68.627     54.688     60.870                                 
 8    1284.09      68.627     54.688     60.870                                 
 9    1331.92      70.588     56.250     62.609                                 
10    1277.93      66.667     59.375     62.810                                 
11    1183.08      68.852     65.625     67.200                                 
12    1254.01      73.684     65.625     69.421                                 
13    1220.34      71.186     65.625     68.293                                 
14    1205.10      77.193     68.750     72.727                                 
15    1159.80      74.138     67.188     70.492                                 
16    1154.58      73.585     60.938     66.667                                 
17    1066.97      72.222     60.938     66.102                                 
18    1096.35      76.000     59.375     66.667                                 
19    1080.09      75.510   

18    1079.45      78.431     62.500     69.565                                 
19    1072.42      76.471     60.938     67.826                                 
20    1130.36      75.000     60.938     67.241                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      80.392   64.062    71.304


Best F-Score   71.304
Baseline       0.000              

✔ Loaded model 'en_core_sci_md'
Created and merged data for 308 total examples
Using 247 train / 61 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model2402_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    1545.99   

 1    1494.70      45.455     23.438     30.928                                 
 2    1388.62      59.615     48.438     53.448                                 
 3    1377.47      67.857     59.375     63.333                                 
 4    1352.62      68.421     60.938     64.463                                 
 5    1322.40      69.355     67.188     68.254                                 
 6    1323.54      70.492     67.188     68.800                                 
 7    1325.43      67.742     65.625     66.667                                 
 8    1195.20      68.254     67.188     67.717                                 
 9    1171.08      66.129     64.062     65.079                                 
10    1179.00      65.672     68.750     67.176                                 
11    1093.51      68.182     70.312     69.231                                 
12    1199.95      69.231     70.312     69.767                                 
13    1173.48      70.492   

12    1228.52      73.077     59.375     65.517                                 
13    1176.59      72.222     60.938     66.102                                 
14    1143.49      73.585     60.938     66.667                                 
15    1131.78      74.074     62.500     67.797                                 
16    1125.37      72.222     60.938     66.102                                 
17    1058.72      70.370     59.375     64.407                                 
18    1083.66      73.585     60.938     66.667                                 
19    1112.12      70.909     60.938     65.546                                 
20    1094.54      72.727     62.500     67.227                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      75.472   62.500    68.376


Best F-Score   68.376
Baseline       0.000              

✔

✔ Loaded model 'en_core_sci_md'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model238_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2841.63      66.667     51.240     57.944                                 
 2    2541.54      71.000     58.678     64.253                                 
 3    2624.98      74.286     64.463     69.027                                 
 4    2479.19      72.381     62.810     67.257                                 
 5    2270.30      76.000     62.810     68.778                                 
 6    2212.17      78.788     64.463     70.909                                 
 7    2110.47      77.451     65.289     70

 5    2361.31      73.469     59.504     65.753                                 
 6    2303.57      76.842     60.331     67.593                                 
 7    2176.19      75.000     64.463     69.333                                 
 8    2213.70      67.890     61.157     64.348                                 
 9    2150.87      66.972     60.331     63.478                                 
10    2167.82      69.231     59.504     64.000                                 
11    2003.04      70.476     61.157     65.487                                 
12    2087.85      70.588     59.504     64.574                                 
13    1997.76      71.569     60.331     65.471                                 
14    1983.34      71.845     61.157     66.071                                 
15    1937.10      71.154     61.157     65.778                                 
16    1839.76      67.593     60.331     63.755                                 
17    1801.46      68.807   

16    1852.45      77.670     66.116     71.429                                 
17    1763.95      77.885     66.942     72.000                                 
18    1850.99      75.000     66.942     70.742                                 
19    1760.18      74.074     66.116     69.869                                 
20    1793.47      75.472     66.116     70.485                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      79.048   68.595    73.451


Best F-Score   73.451
Baseline       0.000              

✔ Loaded model 'en_core_sci_md'
Created and merged data for 547 total examples
Using 438 train / 109 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model4_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  

✔ Initializing with tok2vec weights
models/tok2vec_rf_sci_model989.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    2887.46      60.606     49.587     54.545                                 
 2    2691.38      64.211     50.413     56.481                                 
 3    2570.09      70.270     64.463     67.241                                 
 4    2378.41      73.333     63.636     68.142                                 
 5    2345.38      71.429     61.983     66.372                                 
 6    2291.42      70.000     63.636     66.667                                 
 7    2229.52      71.963     63.636     67.544                                 
 8    2167.06      72.816     61.983     66.964                                 
 9    2082.87      72.816     61.983     66.964                              

 6    2704.25      61.047     66.456     63.636                                 
 7    2515.93      64.458     67.722     66.049                                 
 8    2408.94      64.848     67.722     66.254                                 
 9    2225.26      62.722     67.089     64.832                                 
10    2307.57      62.209     67.722     64.848                                 
11    2218.61      62.428     68.354     65.257                                 
12    2245.70      61.017     68.354     64.478                                 
13    2187.27      60.227     67.089     63.473                                 
14    2113.84      59.444     67.722     63.314                                 
15    2176.02      59.551     67.089     63.095                                 
16    2209.57      62.011     70.253     65.875                                 
17    2125.54      61.538     70.886     65.882                                 
18    2127.06      61.667   

16    2259.31      62.286     68.987     65.465                                 
17    2172.34      60.819     65.823     63.222                                 
18    2185.19      60.234     65.190     62.614                                 
19    2099.32      60.588     65.190     62.805                                 
20    2158.36      61.628     67.089     64.242                                 

============================= ✨  Results summary =============================

Label         Precision   Recall   F-Score
-----------   ---------   ------   -------
RISK_FACTOR      61.798   69.620    65.476


Best F-Score   65.476
Baseline       0.000              

✔ Loaded model 'en_core_sci_md'
Created and merged data for 625 total examples
Using 500 train / 125 eval (split 20%)
Component: ner | Batch size: compounding | Dropout: 0.2 | Iterations: 20
✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model304_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨

✔ Initializing with tok2vec weights
models/tok2vec_abs_sci_model4_gpu.bin

ℹ Baseline accuracy: 0.000

=========================== ✨  Training the model ===========================

#    Loss       Precision   Recall     F-Score 
--   --------   ---------   --------   --------
 1    3252.78      63.158     37.975     47.431                                 
 2    3145.03      55.556     60.127     57.751                                 
 3    2881.81      58.182     60.759     59.443                                 
 4    2755.29      62.893     63.291     63.091                                 
 5    2683.58      60.241     63.291     61.728                                 
 6    2673.09      61.585     63.924     62.733                                 
 7    2510.47      61.078     64.557     62.769                                 
 8    2501.42      61.538     65.823     63.609                                 
 9    2392.42      62.275     65.823     64.000                           

 8    2481.98      63.636     70.886     67.066                                 
 9    2308.64      61.667     70.253     65.680                                 
10    2331.56      61.667     70.253     65.680                                 
11    2263.43      61.364     68.354     64.671                                 
12    2270.75      59.777     67.722     63.501                                 
13    2240.81      60.112     67.722     63.690                                 
14    2167.25      59.669     68.354     63.717                                 
15    2222.46      60.112     67.722     63.690                                 
16    2221.35      60.894     68.987     64.688                                 
17    2218.76      62.360     70.253     66.071                                 
18    2190.62      60.773     69.620     64.897                                 
19    2090.48      60.773     69.620     64.897                                 
20    2124.58      61.111   

### Save top `md` models

In [ ]:
#TODO

### Package top `md` models

In [ ]:
#TODO

# TODO

Try to use the `teach` recipe!